### PR TITLE
2229: Don't move out of drop type

### LINKED ISSUE #88476: missing rust_2021_incompatible_closure_captures with manual drop and move closure
I tried this code:

```rust
#![warn(rust_2021_incompatible_closure_captures)]
#![allow(unused)]
use std::rc::Rc;

struct Foo(Rc<i32>);

impl Drop for Foo {
    fn drop(self: &mut Foo) {}
}

pub fn foo() {
    let f = Foo(Rc::new(1));
    let x = move || {
        println!("{:?}", f.0);
    };
}
```

This does not issue a warning for `rust_2021_incompatible_closure_captures`, however it fails to compile with `capture_disjoint_fields` enabled:

```rust
error[E0509]: cannot move out of type `Foo`, which implements the `Drop` trait
  --> src/lib.rs:13:13
   |
13 |     let x = move || {
   |             ^^^^^^^
   |             |
   |             cannot move out of here
   |             move out of `f.0` occurs here
14 |         println!("{:?}", f.0);
   |                          ---
   |                          |
   |                          move occurs because `f.0` has type `Rc<i32>`, which does not implement the `Copy` trait
   |                          move occurs due to use in closure
```

I would expect `rust_2021_incompatible_closure_captures` to give a suggestion to add `let _ = &f`.

Found in the 2021 crater run for:
* https://tiny.cc/oaaiuz/reg/stakker_async_await-0.0.1/log.txt
* https://tiny.cc/oaaiuz/reg/mobc-0.7.3/log.txt
* https://tiny.cc/oaaiuz/reg/sodium-rust-2.1.1/log.txt
* https://tiny.cc/oaaiuz/reg/glommio-0.5.1/log.txt
* https://tiny.cc/oaaiuz/reg/vapix-0.1.0/log.txt

I'm uncertain if this is a duplicate of #88444 or #88114, they seem very similar, but this case is specific to implementing Drop.

### Meta

`rustc --version --verbose`:
```
rustc 1.56.0-nightly (5eacec9ec 2021-08-28)
binary: rustc
commit-hash: 5eacec9ec7e112a0de1011519a57c45586d58414
commit-date: 2021-08-28
host: x86_64-apple-darwin
release: 1.56.0-nightly
LLVM version: 13.0.0
```

cc @rust-lang/wg-rfc-2229

### PR DESCRIPTION
**Pull request by arora-aman**
Fixes #88476

r? @nikomatsakis

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

As discussed in our [last triage meeting], this is an interesting semantic question and a tweak to the closure rules. I've written up the [my conclusion](https://hackmd.io/AC5EJng9SDStdsjG2M9cUQ?view#Niko%E2%80%99s-conclusion) in this hackmd, which also includes a number of other details. I wanted to do an FCP to get ✅ from the @rust-lang/lang team that they've seen this change. There is some urgency here owing to the Rust 2021 edition timeline.

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 1]
```diff
@@ -626,7 +633,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             self.tcx.struct_span_lint_hir(
                 lint::builtin::RUST_2021_INCOMPATIBLE_CLOSURE_CAPTURES,
                 closure_hir_id,
-                closure_head_span,
+                 closure_head_span,
```
Comment (nbdd0121) [inline compiler/rustc_typeck/src/check/upvar.rs:636]: I am a bit surprised that rustfmt isn't complaining about this extra space.

Comment (arora-aman) [reply compiler/rustc_typeck/src/check/upvar.rs:636]: I'm surprised `./x.py fmt` made this change

**Comment by camelid** [comment]
@arora-aman Not a big deal at all, but was it intentional that you renamed this PR from `2229: Handle drop_defined_for_type during migration` to `2229: Don't move out of drop type 3d7c237`?

**Comment by arora-aman** [comment]
> @arora-aman Not a big deal at all, but was it intentional that you renamed this PR from `2229: Handle drop_defined_for_type during migration` to `2229: Don't move out of drop type 3d7c237`?

wwops, the commit didn't need to be there. But yes we changed how we handle the issue -- so I updated the title

**Comment by camelid** [comment]
(Also just noting that I think this will need to be backported to beta now that the branch has happened.)

**Comment by scottmcm** [comment]
I was convinced in the triage meeting that there are already enough consequences of `Drop` and `Copy` that it's not weird for things to be trait-dependent here.  And since it only comes up in `move` closures, being more aggressive about moving larger things there seems reasonable to me.

@rfcbot reviewed

**Comment by nikomatsakis** [comment]
@bors r+ p=1

Rust 2021 related

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-09-09.20.2354818/near/252633238)

@rustbot label +beta-accepted

**Pull request merged**